src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java

```diff
@@ -0,0 +1,251 @@
+import org.embulk.spi.Exec;
+import org.slf4j.Logger;
+
+import java.time.Instant;
+
+import java.util.Iterator;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+public abstract class ZendeskNormalServices implements ZendeskService
+{
+    private static final Logger logger = Exec.getLogger(ZendeskNormalServices.class);
+
+    protected ZendeskInputPlugin.PluginTask task;
+
+    private ZendeskRestClient zendeskRestClient;
+
+    protected ZendeskNormalServices(final ZendeskInputPlugin.PluginTask task)
+    {
+        this.task = task;
+    }
+
+    protected abstract String buildURI(int page, long startTime);
+
+    @VisibleForTesting
+    protected ZendeskRestClient getZendeskRestClient()
+    {
+        if (zendeskRestClient == null) {
+            zendeskRestClient = new ZendeskRestClient();
+        }
+        return zendeskRestClient;
+    }
+
+    public TaskReport execute(final int taskIndex, final RecordImporter recordImporter)
+    {
+        TaskReport taskReport = Exec.newTaskReport();
+
+        if (isSupportIncremental()) {
+            importDataForIncremental(task, recordImporter, taskReport);
+        }
+        else {
+            importDataForNonIncremental(task, taskIndex, recordImporter);
+        }
+
+        return taskReport;
+    }
+
+    public JsonNode getData(String path, final int page, final boolean isPreview, final long startTime)
+    {
+        if (path.isEmpty()) {
+            path = buildURI(page, startTime);
+        }
+        try {
+            final String response = getZendeskRestClient().doGet(path, task, isPreview);
+            return ZendeskUtils.parseJsonObject(response);
+        }
+        catch (final Exception e) {
+            throw Throwables.propagate(e);
+        }
+    }
+
+    protected URIBuilder getURIBuilderFromHost()
+    {
+        return ZendeskUtils.getURIBuilder(task.getLoginUrl());
+    }
+
+    private void importDataForIncremental(final ZendeskInputPlugin.PluginTask task, final RecordImporter recordImporter, final TaskReport taskReport)
+    {
+        long startTime = 0;
+        if (!Exec.isPreview() && isSupportIncremental()) {
+            if (task.getStartTime().isPresent()) {
+                startTime = ZendeskDateUtils.getStartTime(task.getStartTime().get());
+            }
+        }
+
+        // For incremental target, we will run in one task but split in multiple threads inside for data deduplication.
+        // Run with incremental will contain duplicated data.
+        ThreadPoolExecutor pool = null;
+        try {
+            final Set<String> knownIds = ConcurrentHashMap.newKeySet();
+            pool = new ThreadPoolExecutor(
+                    10, 100, 60, TimeUnit.SECONDS, new LinkedBlockingQueue<>()
+            );
+
+            long resultEndTime = 0;
```

Comment:
I think we actually don't need that `resultEndTime` local var, just assign `result.get(ZendeskConstants.Field.END_TIME).asLong()` to `startTime` directly at the end.

Reply:
We need to store last end_time to update config_diff 

>  if (!Exec.isPreview()) {
                storeStartTimeForConfigDiff(taskReport, resultEndTime);
            }

Reply:
Actually we can just use `startTime` as a replacement for `resultEndTime`, what if:

```
diff --git a/src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java b/src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java
index 84921a3..6f51f94 100644
--- a/src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java
+++ b/src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java
@@ -92,14 +92,12 @@ public abstract class ZendeskNormalServices implements ZendeskService
                     10, 100, 60, TimeUnit.SECONDS, new LinkedBlockingQueue<>()
             );
 
-            long resultEndTime = 0;
             while (true) {
                 int recordCount = 0;
 
                 // Page argument isn't used in incremental API so we just set it to 0
                 final JsonNode result = getData("", 0, false, startTime);
                 final Iterator<JsonNode> iterator = ZendeskUtils.getListRecords(result, task.getTarget().getJsonName());
-                resultEndTime = result.get(ZendeskConstants.Field.END_TIME).asLong();
 
                 int numberOfRecords = 0;
                 if (result.has(ZendeskConstants.Field.COUNT)) {
@@ -130,16 +128,15 @@ public abstract class ZendeskNormalServices implements ZendeskService
                 }
 
                 logger.info("Fetched '{}' records from start_time '{}'", recordCount, startTime);
+                startTime = result.get(ZendeskConstants.Field.END_TIME).asLong();
 
                 if (numberOfRecords < ZendeskConstants.Misc.MAXIMUM_RECORDS_INCREMENTAL) {
                     break;
                 }
-
-                startTime = resultEndTime;
             }
 
             if (!Exec.isPreview()) {
-                storeStartTimeForConfigDiff(taskReport, resultEndTime);
+                storeStartTimeForConfigDiff(taskReport, startTime);
             }
         }
         finally {
```

Reply:
fixed